Add infra agent config option

Guillaume-Barrier opened the pull request:
### What does this PR do?
Merge core payload flags into a single `core_agent` flag available for users to disable. 

### Motivation
This is going to be used for Error Tracking Standalone where Infrastructure Monitoring is not needed, and makes it easier for users to understand and disable it. Those 4 flags are also disabled if users only want Log Collection through the agent (see [public doc](https://docs.datadoghq.com/logs/guide/how-to-set-up-only-logs/?tab=host)); process is now easier and less error prone.


https://datadoghq.atlassian.net/browse/ERRORT-4748

Guillaume-Barrier committed:
Add infra agent config option

Merge core payload flags into a single infra_agent flag. This is going
to be used for Error Tracking Standalone where Infrastructure Monitoring
is not needed, and makes it easier for users to understand and disable
it.

maycmlee commented on pkg/config/config_template.yaml:
```diff
@@ -1636,6 +1636,14 @@ api_key:
   # disable_realtime_checks: false
 
+  ## @param infra_agent - object - optional
+  ## Enables Infrastructure Monitoring
```
```suggestion
  ## Enables infrastructure monitoring
```

Guillaume-Barrier replied:
Hi! After discussing a bit with Product, we may not want to have that much exposure in template/release notes, so removing those entirely in https://github.com/DataDog/datadog-agent/pull/30137/commits/953c6ba87ede9c0c9d3c6dc6e8a8116db3d50cb3

maycmlee reviewed (approved):
Left a couple of small edits, but approving!

ogaca-dd commented on pkg/config/setup/config.go:
```diff
@@ -1648,6 +1649,17 @@ func podman(config pkgconfigmodel.Setup) {
+func infraAgent(config pkgconfigmodel.Setup) {
+	// Disables metric data submission (including Custom Metrics) so that hosts stop showing up in Datadog.
+	// Used namely for Error Tracking Standalone where Infrastructure Monitoring is not needed.
+	if enabled := os.Getenv("DD_INFRA_AGENT_ENABLE"); enabled == "false" {
```
I may miss something but it is quite unusual to use the env variable. I would expect to have `config.GetBool("infra_agent")` instead.
Also the config option "infra_agent" is also not defined and so don't have a default value.

Guillaume-Barrier committed:
Rename flag to default_payloads

Also remove it from the config template and remove the release note.

hush-hush commented on pkg/config/setup/config.go:
```diff
@@ -1020,6 +1020,10 @@ func InitConfig(config pkgconfigmodel.Setup) {
 	config.BindEnvAndSetDefault("reverse_dns_enrichment.rate_limiter.recovery_interval", time.Duration(0))
+
+	// Default payloads (disabled for Error Tracking Standalone, Logs Collection Only)
+	config.BindEnvAndSetDefault("default_payloads.enabled", true)
```
Could we be more specific in the naming ? What about `infra_payloads` ?

Also, are we planning more setting in this section ? If now what about `infra_payloads_enabled` or something similar ?

carlosroman replied:
I think it was more following the pattern other configs have. Agree that it should be something with `infra` in the name (I'm still voting for something like `infra_monitoring`)

carlosroman committed:
Adding check that default payloads is enabled (#30600)

Guillaume-Barrier committed:
Rename flag to core_agent

Name seems more meaningful and can be reused more easily in the future.

carlosroman commented on pkg/config/utils/miscellaneous_test.go:
```diff
@@ -0,0 +1,66 @@
+	"github.com/DataDog/datadog-agent/pkg/config/model"
+)
+
+func TestIsDefaultPayloadsEnabled(t *testing.T) {
```
```suggestion
func TestIsCoreAgentEnabled(t *testing.T) {
```

Guillaume-Barrier committed:
Update pkg/config/utils/miscellaneous_test.go

Co-authored-by: Carlos <carlos.roman@datadoghq.com>

Guillaume-Barrier committed:
Merge branch 'main' into guillaume.barrier/add-infra-agent-config-option

Guillaume-Barrier committed:
Update names

pgimalac commented on pkg/config/setup/config.go:
```diff
@@ -2357,6 +2361,17 @@ func sanitizeExternalMetricsProviderChunkSize(config pkgconfigmodel.Config) {
+	// Disables metric data submission (including Custom Metrics) so that hosts stop showing up in Datadog.
+	// Used namely for Error Tracking Standalone where it is not needed.
+	if !config.GetBool("core_agent.enabled") {
+		config.BindEnvAndSetDefault("enable_payloads.events", false)
+		config.BindEnvAndSetDefault("enable_payloads.series", false)
+		config.BindEnvAndSetDefault("enable_payloads.service_checks", false)
+		config.BindEnvAndSetDefault("enable_payloads.sketches", false)
```
This redeclares the config key with a different default rather than override their value, I believe this is the correct way to do it
```suggestion
		config.Set("enable_payloads.events", false, pkgconfigmodel.SourceAgentRuntime)
		config.Set("enable_payloads.series", false, pkgconfigmodel.SourceAgentRuntime)
		config.Set("enable_payloads.service_checks", false, pkgconfigmodel.SourceAgentRuntime)
		config.Set("enable_payloads.sketches", false, pkgconfigmodel.SourceAgentRuntime)
```

Guillaume-Barrier committed:
Update pkg/config/setup/config.go

Co-authored-by: Pierre Gimalac <pierre.gimalac@datadoghq.com>

Guillaume-Barrier commented on comp/trace/config/setup.go:
```diff
@@ -194,7 +194,8 @@ func applyDatadogConfig(c *config.AgentConfig, core corecompcfg.Component) error
 		c.SkipSSLValidation = core.GetBool("skip_ssl_validation")
 	}
 	if core.IsSet("apm_config.enabled") {
```
It looks consistant with the rest like this, and this way no one finds it weird and asks questions as well :)

Guillaume-Barrier committed:
Merge branch 'main' into guillaume.barrier/add-infra-agent-config-option

Guillaume-Barrier committed:
Merge branch 'main' into guillaume.barrier/add-infra-agent-config-option

Guillaume-Barrier committed:
Allow APM enablement if no core but ETS

If the core agent is disabled but the host is set as Error Tracking
Standalone, we need to enable the trace agent.

Guillaume-Barrier committed:
Allow APM enablement if no APM but ETS

Users don't have to explicitely enable APM to get ETS, yet they need
core_agent if they want APM without ETS

GianlucaBortoli committed:
Merge branch 'main' into guillaume.barrier/add-infra-agent-config-option

GianlucaBortoli commented:
/merge

GianlucaBortoli commented:
/merge

ichinaski commented on comp/trace/config/setup.go:
```diff
@@ -194,7 +194,9 @@ func applyDatadogConfig(c *config.AgentConfig, core corecompcfg.Component) error
-		c.Enabled = core.GetBool("apm_config.enabled")
+		isAPMEnabled := (core.GetBool("apm_config.enabled") && utils.IsCoreAgentEnabled(core)) ||
+			core.GetBool("apm_config.error_tracking_standalone.enabled")
+		c.Enabled = isAPMEnabled
```
Why does the trace-agent now require this extra logic to determine if it's enabled when an explicit setting exists? 🤔 
Isn't `apm_config.enabled` a requisite for ETS?

ichinaski replied:
Thank you. The doc doesn't mention that `error_tracking_standalone.enabled` is nested under `apm_config`.  If the customer is configuring ETS under `apm_config`, I would expect them to also `enable` it? 🤔

GianlucaBortoli replied:
> The doc doesn't mention that error_tracking_standalone.enabled is nested under apm_config

That's missing in the doc, good catch. Fixing it there as it's supposed to be under `apm_config`

> If the customer is configuring ETS under apm_config, I would expect them to also enable it?

No, the goal was to have the minimum number of options that need to be enabled by the user. The best agreement we found w/ PMs on the agent side was to have only  `apm_config.error_tracking_standalone.enabled` and not require `apm_config.enabled` to have fewer settings to change to enable the functionality.

ichinaski commented on pkg/config/setup/config.go:
```diff
@@ -2384,6 +2388,17 @@ func sanitizeExternalMetricsProviderChunkSize(config pkgconfigmodel.Config) {
+		config.Set("enable_payloads.series", false, pkgconfigmodel.SourceAgentRuntime)
+		config.Set("enable_payloads.service_checks", false, pkgconfigmodel.SourceAgentRuntime)
+		config.Set("enable_payloads.sketches", false, pkgconfigmodel.SourceAgentRuntime)
+	}
```
what happens if the customer has defined these settings manually?

ichinaski commented on pkg/config/utils/miscellaneous.go:
```diff
@@ -31,3 +31,21 @@ func SetLogLevel(level string, config pkgconfigmodel.Writer, source pkgconfigmod
 }
+
+// IsCoreAgentEnabled checks if the Agent is able to send the payloads it and other Agents need to function with
+func IsCoreAgentEnabled(cfg pkgconfigmodel.Reader) bool {
```
nit: IMO this function name is bit deceptive, I would have expected only `core_agent.enabled` to be relevant for it's resolution

GianlucaBortoli committed:
Merge branch 'main' into guillaume.barrier/add-infra-agent-config-option

ichinaski reviewed (commented):
I think the duplicated boolean logic needs more/better documentation on why it's that way, or just rely on the existing `apm_config.enabled`.

I also believe the `core_agent.enabled` and the `enable_payloads.*` logic needs more comments or linked documentation on what it does. It's hard to reason about this code without reading the doc linked in the review comments.

Could you also add QA label or QA instructions? 🙏

GianlucaBortoli commented:
/merge

GianlucaBortoli committed:
Unify check for APM enabled into a helper function
